aws-cpp-sdk-transfer/source/transfer/UploadFileRequest.cpp

```diff
@@ -91,56 +90,125 @@ m_getObjectRetries(0),
+                                     const Aws::String& bucketName,
+                                     const Aws::String& keyName,
+                                     const Aws::String& contentType,
                                      const Aws::Map<Aws::String, Aws::String>& metadata,
                                      const std::shared_ptr<Aws::S3::S3Client>& s3Client,
                                      const std::shared_ptr<Aws::Utils::Threading::BlockingExecutor>& executor,
                                      bool createBucket,
                                      bool doConsistencyChecks) :
-UploadFileRequest(fileName, bucketName, keyName, contentType, Aws::Map<Aws::String, Aws::String>(metadata), s3Client, executor, createBucket, doConsistencyChecks)
+UploadFileRequest(fileStream, bucketName, keyName, contentType, Aws::Map<Aws::String, Aws::String>(metadata), s3Client, executor, createBucket, doConsistencyChecks)
+{
+}
+    
+UploadFileRequest::UploadFileRequest(std::shared_ptr<Aws::IOStream> fileStream,
+                                     const Aws::String& bucketName,
+                                     const Aws::String& keyName,
+                                     const Aws::String& contentType,
+                                     const std::shared_ptr<Aws::S3::S3Client>& s3Client,
+                                     const std::shared_ptr<Aws::Utils::Threading::BlockingExecutor>& executor,
+                                     bool createBucket,
+                                     bool doConsistencyChecks) :
+UploadFileRequest(fileStream, bucketName, keyName, contentType, Aws::Map<Aws::String, Aws::String>{}, s3Client, executor, createBucket, doConsistencyChecks)
 {
 }
 
 UploadFileRequest::~UploadFileRequest()
 {
     m_executor->WaitForCompletion();
-    m_fileStream.close();
+    if (auto fileStream = std::dynamic_pointer_cast<Aws::FStream>(m_inputStream))
+    {
+        fileStream->close();
+    }
 }
-
-bool UploadFileRequest::CreateBucket()
+    
+/* Check input stream, find file/stream size, and calculate total parts */
+void UploadFileRequest::Init()
+{
+    if (m_inputStream->good())
+    {
+        // If input stream is a file stream, make sure it's open
+        if (auto fileStream = std::dynamic_pointer_cast<Aws::FStream>(m_inputStream))
+        {
+            if (!fileStream->is_open())
+            {
+                std::cout << "fail" << std::endl;
+                CompletionFailure("Failed to open file.");
+                return;
+            }
+        }
+        
+        m_inputStream->seekg(0, std::ios_base::end);
+        SetFileSize(static_cast<uint64_t>(m_inputStream->tellg()));
+        m_bytesRemaining = GetFileSize();
+        m_inputStream->seekg(0);
+    }
+    else
+    {
+        std::cout << "BAD" << std::endl;
+        CompletionFailure("Failed to open file.");
+        return;
+    }
+    
+    if (GetFileSize())

```

Comment:
I think it makes sense to upload empty files to S3 (e.g. empty logs), so I've added the functionality and a test for it.